File: server/src/entities/Task.ts
import {
    Column,
    CreateDateColumn,
    Entity,
    JoinColumn,
    ManyToOne,
    ObjectId,
    ObjectIdColumn,
    OneToMany,
    UpdateDateColumn
} from "typeorm"
import { Comment } from "./Comment"
import { TaskPriority } from "./PriorityEnum"
import { Project } from "./Project"
import { TaskStatus } from "./StatusEnum"
import { Tag } from "./Tag"
import { User } from "./User"

@Entity("taskr_tasks")
export class Task {
    @ObjectIdColumn()
    _id: ObjectId

    @Column({ unique: true, nullable: false })
    title: string

    @Column({ nullable: false, type: "text" })
    description: string

    @Column({ nullable: false })
    projectId: ObjectId

    @ManyToOne(() => Project, (project) => project.tasks, { eager: false })
    @JoinColumn()
    project: Project

    @Column({ nullable: false })
    creatorId: ObjectId

    @ManyToOne(() => User, (user) => user.tasks, { eager: false })
    @JoinColumn()
    creator: User

    @Column({ type: "enum", enum: TaskStatus, default: TaskStatus.PLANNED })
    status: TaskStatus

    @Column({ type: "enum", enum: TaskPriority, default: TaskPriority.LOW })
    priority: TaskPriority

    @Column({ nullable: false, type: "timestamp" })
    dueDate: Date

    @OneToMany(() => Comment, (comment) => comment.task)
    comments?: Comment[]

    @Column({ nullable: true })
    assigneeId: ObjectId

    @ManyToOne(() => User, (user) => user.assignedTasks, { eager: false })
    @JoinColumn()
    assignee?: User

    @Column({ nullable: true, type: "array" })
    tags?: Tag[]

    @CreateDateColumn({ type: "timestamp", default: () => "CURRENT_TIMESTAMP" })
    createdAt: Date

    @UpdateDateColumn({ type: "timestamp", default: () => "CURRENT_TIMESTAMP", onUpdate: "CURRENT_TIMESTAMP" })
    updatedAt: Date
}

File: server/src/main/java/com/barataribeiro/taskr/task/Task.java
package com.barataribeiro.taskr.task;

import com.barataribeiro.taskr.comment.Comment;
import com.barataribeiro.taskr.project.Project;
import com.barataribeiro.taskr.task.enums.TaskPriority;
import com.barataribeiro.taskr.task.enums.TaskStatus;
import com.barataribeiro.taskr.user.User;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.io.Serial;
import java.io.Serializable;
import java.time.Instant;
import java.time.LocalDateTime;
import java.util.LinkedHashSet;
import java.util.Set;

@Builder
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@ToString
@Entity
@Table(name = "tb_tasks", indexes = {
        @Index(name = "idx_task_id", columnList = "id")
}, uniqueConstraints = {
        @UniqueConstraint(name = "uc_task_title", columnNames = {"title"})
})
public class Task implements Serializable {
    @Serial
    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(updatable = false, nullable = false, unique = true)
    private Long id;

    @Column(name = "title", nullable = false)
    private String title;

    @Lob
    @Column(name = "description", nullable = false)
    private String description;

    @Column(name = "due_date")
    private LocalDateTime dueDate;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private TaskStatus status = TaskStatus.TO_DO;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(name = "priority", nullable = false)
    private TaskPriority priority = TaskPriority.LOW;

    @Builder.Default
    @Column(name = "position", nullable = false)
    private Integer position = 1;

    // Timestamps

    @Column(updatable = false)
    @CreationTimestamp
    private Instant createdAt;

    @UpdateTimestamp
    private Instant updatedAt;

    // Relationships

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id", nullable = false)
    @ToString.Exclude
    @JsonIgnore
    private Project project;

    @Builder.Default
    @ManyToMany(fetch = FetchType.LAZY, cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(name = "task_assignee", joinColumns = @JoinColumn(name = "task_id"),
               inverseJoinColumns = @JoinColumn(name = "user_id"))
    @ToString.Exclude
    @JsonIgnore
    private Set<User> assignees = new LinkedHashSet<>();

    @Builder.Default
    @OneToMany(mappedBy = "task", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    @ToString.Exclude
    @JsonIgnore
    private Set<Comment> comments = new LinkedHashSet<>();


    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .append(getId())
                .append(getTitle())
                .append(getDueDate())
                .append(getStatus())
                .append(getPriority())
                .append(getPosition())
                .toHashCode();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;

        if (!(o instanceof Task task)) return false;

        return new EqualsBuilder()
                .append(getId(), task.getId())
                .append(getTitle(), task.getTitle())
                .append(getDueDate(), task.getDueDate())
                .append(getStatus(), task.getStatus())
                .append(getPriority(), task.getPriority())
                .append(getPosition(), task.getPosition())
                .isEquals();
    }
}
